### ISSUE TITLE
Text search is not returning results of custom fields because of strange encoding

### ISSUE THREAD

**Issue by shaneonabike** [10up/ElasticPress#2333]
**Describe the bug**
Basically, when searching for text within several fields, including the body and title the results appearing only match if the word exists in the body and title. I can see that the fields are being properly added to the search query so I'm not sure why it's not working. I could be wrong on the encoding issue (see below) but through debugging I can see a few strange things.

**Expected behavior**
Searching for a word like _Jesse_ would show results for all matching results that occur in title, body, and custom fields.

**Environment information**
`
### wp-core ###

version: 5.8
site_language: en_CA
user_language: en_CA
timezone: America/Toronto
environment_type: production

### wp-plugins-active (48) ###

ACF Content Analysis for Yoast SEO: version: 3.0.1, author: Thomas Kräftner, ViktorFroberg, marol87, pekz0r, angrycreative, Team Yoast, Auto-updates disabled
Advanced Custom Fields: Star Rating Field: version: 1.0.2, author: Kevin Ruscoe, Auto-updates disabled
Advanced Custom Fields Multilingual: version: 1.8.2, author: OnTheGoSystems, Auto-updates disabled
Advanced Custom Fields PRO: version: 5.9.9, author: Delicious Brains, Auto-updates disabled
Caldera Custom Fields: version: 2.2.1, author: Caldera Labs, Auto-updates disabled
Caldera Forms: version: 1.9.4, author: Caldera Forms, Auto-updates disabled
Caldera Forms Anti Spam: version: 0.4, author: Yoohoo Plugins, Auto-updates disabled
CF CiviCRM: version: 1.0.5, author: Andrei Mondoc, Auto-updates disabled
CiviCRM: version: 5.40.3, author: CiviCRM LLC, Auto-updates disabled
CiviCRM Admin Utilities: version: 0.9, author: Christian Wach, Auto-updates disabled
CiviCRM WordPress Member Sync: version: 0.5.1, author: Christian Wach, Auto-updates disabled
CP Archive Generator: version: 1.0.0, author: Bees on a Bike, Auto-updates disabled
Crop Thumbnails: version: 1.3.1, author: Volkmar Kantor, Auto-updates disabled
Dashboard Welcome for Elementor: version: 1.0.5, author: IdeaBox Creations, Auto-updates disabled
Debug Bar: version: 1.1.2, author: wordpressdotorg, Auto-updates disabled
Debug Bar ElasticPress: version: 2.1.0, author: 10up, Auto-updates disabled
DynamicConditions: version: 1.5.1, author: RTO GmbH, Auto-updates disabled
ElasticPress: version: 3.6.1, author: 10up (latest version: 3.6.2), Auto-updates disabled
Ele Custom Skin: version: 3.1.2, author: Dudaster.com, Auto-updates disabled
Elementor: version: 3.4.2, author: Elementor.com, Auto-updates disabled
Elementor Pro: version: 3.3.8, author: Elementor.com, Auto-updates disabled
Email Address Encoder: version: 1.0.22, author: Till Krüss, Auto-updates disabled
Full Screen Menu for Elementor: version: 1.0.3, author: Solitweb, Auto-updates disabled
Imagify: version: 1.10, author: WP Media, Auto-updates disabled
Members: version: 3.1.5, author: MemberPress, Auto-updates disabled
Ninja Forms: version: 3.5.7, author: Saturday Drive, Auto-updates disabled
Ninja Forms - Conditional Logic: version: 3.1, author: The WP Ninjas, Auto-updates disabled
Ninja Forms - Layout & Styles: version: 3.0.29, author: The WP Ninjas, Auto-updates disabled
Ninja Forms - Multi-Part Forms: version: 3.0.26, author: The WP Ninjas, Auto-updates disabled
Ninja Forms - Post Creation: version: 3.0.10, author: The WP Ninjas, Auto-updates disabled
Ninja Forms Multilingual: version: 0.2.0, author: OnTheGoSystems, Auto-updates disabled
Oasis Workflow: version: 5.7, author: Nugget Solutions Inc., Auto-updates disabled
Official Mad Mimi Signup Forms: version: 1.7.1, author: Mad Mimi, LLC, Auto-updates disabled
Redis Object Cache: version: 2.0.21, author: Till Krüss, Auto-updates disabled
Safe SVG: version: 1.9.9, author: Daryll Doyle, Auto-updates disabled
Sticky Header Effects for Elementor: version: 1.4.7, author: Rwattner, Auto-updates disabled
UpdraftPlus - Backup/Restore: version: 1.16.60, author: UpdraftPlus.Com, DavidAnderson, Auto-updates disabled
Wordfence Login Security: version: 1.0.8, author: Wordfence, Auto-updates disabled
WP fail2ban: version: 4.3.0.9, author: Charles Lecklider, Auto-updates disabled
WPML ElasticPress Integration: version: 1.0.0, author: OnTheGoSystems, Auto-updates disabled
WPML Media: version: 2.6.4, author: OnTheGoSystems, Auto-updates disabled
WPML Multilingual CMS: version: 4.4.10, author: OnTheGoSystems, Auto-updates disabled
WPML SEO: version: 2.0.1, author: OnTheGoSystems, Auto-updates disabled
WPML String Translation: version: 3.1.8, author: OnTheGoSystems, Auto-updates disabled
WPML Translation Management: version: 2.10.6, author: OnTheGoSystems, Auto-updates disabled
WPS Hide Login: version: 1.8.5, author: WPServeur, NicolasKulka, wpformation, Auto-updates disabled
Yoast Duplicate Post: version: 4.1.2, author: Enrico Battocchi & Team Yoast, Auto-updates disabled
Yoast SEO: version: 17.0, author: Team Yoast, Auto-updates disabled
`

**Additional context**
From the Debug Bar when I search for _Jesse Freeston_ I'm getting the following query below. It is adding the proper fields and normally it should find the name within meta.film_castcrew.value.

```
{
    "from": 0,
    "size": 16,
    "sort": [
        {
            "meta.film_prod_year.raw": {
                "order": "desc"
            }
        }
    ],
    "query": {
        "function_score": {
            "query": {
                "bool": {
                    "should": [
                        {
                            "bool": {
                                "must": [
                                    {
                                        "bool": {
                                            "should": [
                                                {
                                                    "multi_match": {
                                                        "query": "Jesse Freeston",
                                                        "type": "phrase",
                                                        "fields": [
                                                            "post_title^80",
                                                            "post_content^78",
                                                            "terms.languages.name^57",
                                                            "terms.countries.name^58",
                                                            "terms.film_categories.name^86",
                                                            "meta.film_title.value^65",
                                                            "meta.tagline.value^68",
                                                            "meta.film_castcrew.value^67",
                                                            "meta.film_prod_year.value^67",
                                                            "meta.film_formats.value^68",
                                                            "meta.artists.value^29",
                                                            "meta.film_vod.value^45"
                                                        ],
                                                        "boost": 3
                                                    }
                                                },
                                                {
                                                    "multi_match": {
                                                        "query": "Jesse Freeston",
                                                        "fields": [
                                                            "post_title^80",
                                                            "post_content^78",
                                                            "terms.languages.name^57",
                                                            "terms.countries.name^58",
                                                            "terms.film_categories.name^86",
                                                            "meta.film_title.value^65",
                                                            "meta.tagline.value^68",
                                                            "meta.film_castcrew.value^67",
                                                            "meta.film_prod_year.value^67",
                                                            "meta.film_formats.value^68",
                                                            "meta.artists.value^29",
                                                            "meta.film_vod.value^45"
                                                        ],
                                                        "type": "phrase",
                                                        "slop": 5
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                ],
                                "filter": [
                                    {
                                        "match": {
                                            "post_type.raw": "film"
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
            },
            "functions": [
                {
                    "exp": {
                        "post_date_gmt": {
                            "scale": "14d",
                            "decay": 0.25,
                            "offset": "7d"
                        }
                    }
                },
                {
                    "weight": 0.001
                }
            ],
            "score_mode": "sum",
            "boost_mode": "multiply"
        }
    },
    "post_filter": {
        "bool": {
            "must": [
                {
                    "bool": {
                        "must": [
                            {
                                "bool": {
                                    "should": {
                                        "bool": {
                                            "should": [
                                                {
                                                    "terms": {
                                                        "meta.sticky_post.raw": [
                                                            0
                                                        ]
                                                    }
                                                },
                                                {
                                                    "bool": {
                                                        "must_not": [
                                                            {
                                                                "exists": {
                                                                    "field": "meta.sticky_post"
                                                                }
                                                            }
                                                        ]
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                }
                            }
                        ]
                    }
                },
                {
                    "terms": {
                        "post_type.raw": [
                            "film"
                        ]
                    }
                },
                {
                    "terms": {
                        "post_status": [
                            "publish",
                            "acf-disabled",
                            "dp-rewrite-republish"
                        ]
                    }
                },
                {
                    "term": {
                        "post_lang": "en"
                    }
                }
            ]
        }
    },
    "aggs": {
        "terms": {
            "filter": {
                "bool": {
                    "must": [
                        {
                            "bool": {
                                "must": [
                                    {
                                        "bool": {
                                            "should": {
                                                "bool": {
                                                    "should": [
                                                        {
                                                            "terms": {
                                                                "meta.sticky_post.raw": [
                                                                    0
                                                                ]
                                                            }
                                                        },
                                                        {
                                                            "bool": {
                                                                "must_not": [
                                                                    {
                                                                        "exists": {
                                                                            "field": "meta.sticky_post"
                                                                        }
                                                                    }
                                                                ]
                                                            }
                                                        }
                                                    ]
                                                }
                                            }
                                        }
                                    }
                                ]
                            }
                        },
                        {
                            "terms": {
                                "post_type.raw": [
                                    "film"
                                ]
                            }
                        },
                        {
                            "terms": {
                                "post_status": [
                                    "publish",
                                    "acf-disabled",
                                    "dp-rewrite-republish"
                                ]
                            }
                        }
                    ]
                }
            },
            "aggs": {
                "category": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.category.slug"
                    }
                },
                "post_tag": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.post_tag.slug"
                    }
                },
                "post_format": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.post_format.slug"
                    }
                },
                "post_status": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.post_status.slug"
                    }
                },
                "languages": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.languages.slug"
                    }
                },
                "countries": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.countries.slug"
                    }
                },
                "film_categories": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.film_categories.slug"
                    }
                },
                "master_copies": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.master_copies.slug"
                    }
                },
                "pricing_grid": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.pricing_grid.slug"
                    }
                },
                "press_types": {
                    "terms": {
                        "size": 10000,
                        "field": "terms.press_types.slug"
                    }
                }
            }
        }
    }
}
```

The artists are stored like the following (retrieved by searching for Nettie who's name is in the body of the content type).
```
"film_castcrew_0_cast_name": [
                            {
                                "value": "Abraham Lifshitz",
                                "raw": "Abraham Lifshitz",
                                "boolean": false,
                                "date": "1970-01-01",
                                "datetime": "1970-01-01 00:00:01",
                                "time": "00:00:01"
                            }
                        ],
                        "film_castcrew_1_cast_name": [
                            {
                                "value": "Jesse Freestonu00a0andu00a0Diego Briceno",
                                "raw": "Jesse Freestonu00a0andu00a0Diego Briceno",
                                "boolean": false,
                                "date": "1970-01-01",
                                "datetime": "1970-01-01 00:00:01",
                                "time": "00:00:01"
                            }
                        ],
                        "film_castcrew_2_cast_name": [
                            {
                                "value": "Renu00e9 Portillo",
                                "raw": "Renu00e9 Portillo",
                                "boolean": false,
                                "date": "1970-01-01",
                                "datetime": "1970-01-01 00:00:01",
                                "time": "00:00:01"
                            }
                        ],
                        "film_castcrew_3_cast_name": [
                            {
                                "value": "Billy Buntin",
                                "raw": "Billy Buntin",
                                "boolean": false,
                                "date": "1970-01-01",
                                "datetime": "1970-01-01 00:00:01",
                                "time": "00:00:01"
                            }
                        ],
                        "film_castcrew_4_cast_name": [
                            {
                                "value": "Stefan Christoff",
                                "raw": "Stefan Christoff",
                                "boolean": false,
                                "date": "1970-01-01",
                                "datetime": "1970-01-01 00:00:01",
                                "time": "00:00:01"
                            }
                        ]
```

Funny enough I looked in the ACF field and it would appear there was a ```&nbsp;``` pasted into the input field. This is then (as shown above) encoded to 00a0? Also, why is René resulting in ```Renu00e9``` will this affect a search for that name as well?

**Comment by Rahmon** [comment]
Hi @shaneonabike 

Looking at the query you paste, you're searching for `"meta.film_castcrew.value^67"` but the meta-field names are `film_castcrew_0_cast_name`, `film_castcrew_1_cast_name` and so on. So you need to make sure you're indexing these meta fields and searching through them.

You can try to use the [ElasticPress Labs](https://github.com/10up/ElasticPressLabs). This plugin has a developer focused interface to commonly ElasticPress plugin issues. One of the features is the Meta Key Pattern where you can have more control over the metadata indexed.

For example, if you add the pattern `/film_castcrew_/|67`, it will index all metadata that starts with `film_castcrew_` and will apply `67` as weight in the search.

**Comment by shaneonabike** [comment]
Hey I wasn't awae of the ElasticPess Labs that's awesome!

Actually I apologize for not responding to this earlier. I finally discovered the issue. It was indexing all these properly but in terms of search it has been setup to search for ```film_castcrew``` only. It's actually because this is a ACF Repeater field and so there are a number of (unknown) fields being added.

What I did instead is to use the hook for the meta fields, and then concatenated all the ```film_cast_crew_[0-9]_cast_name``` together into one field and dumped that into ```field_castcrew```. This is working magic now and searching perfectly!

**Comment by shaneonabike** [comment]
I had one other question related to this though. Normally, if you do a normal text search is that only performed on the content and title fields? (Meaning using the ```s=blahblah``` variable). I noticed that unless I use a filter specific for taxonomy terms they are never double checked. I think the only solution that I have found is to also dump these into a field that writes out the names in case a user types ```Action``` or something like that in the search field. When they use the filter it works perfectly!

Just wanted to see if I was missing something or that is the normal behaviour?
